File: 20 valid parentheses/ValidParentheses.java

public class ValidParentheses {
    /**
     * Time O(n) Space O(n)
     * We use stack to save all left parentheses and when we meet right parentheses,
     * we check the element at the top of that stack.
     * At last, we check if there is element left in the stack.
     * 
     */
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '[' || c == '{' || c == '(') {
                stack.push(c);
            }
            if (c == ']') {
                if (stack.isEmpty() || stack.peek() != '[') {
                    return false;
                } else {
                    stack.pop();
                }
            }
            if (c == '}') {
                if (stack.isEmpty() || stack.peek() != '{') {
                    return false;
                } else {
                    stack.pop();
                }
            }
            if (c == ')') {
                if (stack.isEmpty() || stack.peek() != '(') {
                    return false;
                } else {
                    stack.pop();
                }
            }
        }
        return stack.isEmpty();
    }
}

File: 20 valid parentheses/ValidParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        
        for i in range(len(s)):
            if s[i] == '[' or s[i] == '(' or s[i] == '{':
                stack.append(s[i])
            elif s[i] == ']':
                if not stack or stack[-1] != '[':
                    return False
                stack.pop()
            elif s[i] == '}':
                if not stack or stack[-1] != '{':
                    return False
                stack.pop()
            elif s[i] == ')':
                if not stack or stack[-1] != '(':
                    return False
                stack.pop()
        
        return not stack